src/sentry/search/snuba/executors.py

```diff
@@ -0,0 +1,667 @@
+            # query has.
+
+            # To do this, we get a sample of groups matching the snuba side of
+            # the query, and see how many of those pass the post-filter in
+            # postgres. This should give us an estimate of the total number of
+            # snuba matches that will be overall matches, which we can use to
+            # get an estimate for X-Hits.
+
+            # The sampling is not simple random sampling. It will return *all*
+            # matching groups if there are less than N groups matching the
+            # query, or it will return a random, deterministic subset of N of
+            # the groups if there are more than N overall matches. This means
+            # that the "estimate" is actually an accurate result when there are
+            # less than N matching groups.
+
+            # The number of samples required to achieve a certain error bound
+            # with a certain confidence interval can be calculated from a
+            # rearrangement of the normal approximation (Wald) confidence
+            # interval formula:
+            #
+            # https://en.wikipedia.org/wiki/Binomial_proportion_confidence_interval
+            #
+            # Effectively if we want the estimate to be within +/- 10% of the
+            # real value with 95% confidence, we would need (1.96^2 * p*(1-p))
+            # / 0.1^2 samples. With a starting assumption of p=0.5 (this
+            # requires the most samples) we would need 96 samples to achieve
+            # +/-10% @ 95% confidence.
+
+            sample_size = options.get("snuba.search.hits-sample-size")
+            snuba_groups, snuba_total = self.snuba_search(
+                start=start,
+                end=end,
+                project_ids=[p.id for p in projects],
+                environment_ids=environments and [environment.id for environment in environments],
+                sort_field=sort_field,
+                limit=sample_size,
+                offset=0,
+                get_sample=True,
+                search_filters=search_filters,
+            )
+            snuba_count = len(snuba_groups)
+            if snuba_count == 0:
+                return self.EMPTY_RESULT
+            else:
+                filtered_count = group_queryset.filter(
+                    id__in=[gid for gid, _ in snuba_groups]
+                ).count()
+
+                hit_ratio = filtered_count / float(snuba_count)
+                hits = int(hit_ratio * snuba_total)
+
+        # Do smaller searches in chunks until we have enough results
+        # to answer the query (or hit the end of possible results). We do
+        # this because a common case for search is to return 100 groups
+        # sorted by `last_seen`, and we want to avoid returning all of
+        # a project's groups and then post-sorting them all in Postgres
+        # when typically the first N results will do.
+        while (time.time() - time_start) < max_time:
```

Comment:
Ok, now, you should be able to see that there is a lot of common logic between the two implementations of AbstractQueryExecutor. It seems to me these are:
- some parameters preparation (start/end/retention)
- chunked query (basically from line 386 to 414
- preparing the paginator to return at the end of this method
There are two ways to proceed to rationalize this:
1) you build a template method (see the comment about snuba_search). Which means, the parent class implement query and the subclasses implement the part that is not in common
2) as an alternative, you keep query in the subclasses and put the common parts into protected methods in the parent class.

your call on this. Either can work if meaningful methods are defined.